### PR TITLE
Generic OAuth: Add support for requiring role

### PR DESCRIPTION
**Pull request by burakksglu**
**What this PR does / why we need it**:

This PR adds a 'login restriction by role' functionality for Generic Auth role mapping. This way if "required_role" is enabled from config, we can allow login only for 'Admin', 'Editor' and 'Viewer' roles. If it is disabled "auto_assign_org_role" is still active and works.

**Which issue(s) this PR fixes**:

This PR fixes the issue of the entry of those who do not have a role. For example, if a user from Keycloak has no role for Grafana client and he tries to login the client anyway, he can still login with some role. What we want is in that situation he shouldn't be able to login that Grafana client. 

Therefore this feature adds a functionality for those who wants to restrict login by role for Generic Auth.

**Special notes for your reviewer**:

We tested this feature on a local Keycloak server and it works as intended. 

**If 'role_required' is true from sample.ini config file:**

_1- User has a Grafana role from Keycloak ( Admin/Editor/Viewer ) he can login to the client.
2- User doesn't have a Grafana role from Keycloak but has something else ( Guest,user,etc. ) he can't login to the client.
3- User doesn't have ANY role from Keycloak, he can't login to the client._ 

**If 'role_required' is false from sample.ini config file:**

_Login works as previous, if auto_assign is enable it assigns corresponding role for the user and let's the user in to client._

Since this is my first PR to Grafana community and I am not a Go expert, I do need a help on documentation and other necessary actions for this PR. 

Thanks!

### CODE REVIEW

**Comment by chelnak** [comment]
It would be cool if we could ask Grafana to validate custom roles.. for example, if my IdP is used for many other things, having `Viewer/Editor/Admin` could be confusing.. so you might want `Grafana_User` for example.. which maps to Viewer..

Is this something that is possible with the current setup?

File: conf/sample.ini
Comment (aknuds1) [inline conf/sample.ini]: Should have the default value here shouldn't it:
```suggestion
;role_required = false
```

File: pkg/login/social/generic_oauth.go
Comment (aknuds1) [inline pkg/login/social/generic_oauth.go]:
```suggestion
	s.log.Debug("Checking if role is allowed", "role", role, "s.roleRequired", s.roleRequired)
```

File: pkg/login/social/generic_oauth.go
Comment (aknuds1) [inline pkg/login/social/generic_oauth.go]:
```suggestion
	}

```

File: pkg/login/social/generic_oauth.go
Comment (aknuds1) [inline pkg/login/social/generic_oauth.go]:
```suggestion

	return false
```

**Comment by burakksglu** [comment]
@chelnak ;

In our configuration it doesn't map to custom role. We only pass the current supported roles by default from Keycloak to Grafana which are 'Admin, 'Editor' and 'Viewer'. We have client(application) based roles to overcome this issue. When we create a new client we also create the roles in it. ('Admin', 'Editor', 'Viewer') 

It would be nice to have what you say though. So we could create a realm role with the prefix. But you can still overcome this on IdP I guess.

**Review by aknuds1 — CHANGES_REQUESTED** [review]
Thanks for your PR, I made some suggestions for improvement. I will need additional feedback though, from team mates with more experience in OAuth.

**Comment by burakksglu** [comment]
@aknuds1 Should I wait for reviewers or commit the changes you've requested?

**Comment by burakksglu** [comment]
I've commited change requests by @aknuds1.
Sorry to bother you guys but did you have any chance to look at this PR? Any updates or thoughts? 

CC: @torkelo @daniellee @bergquist @papagian @marefr

**Review by aknuds1 — APPROVED** [review]
LGTM after the changes. @marefr did we come to a conclusion on merging this?

**Comment by marefr** [comment]
We're going to discuss it today.

**Comment by aknuds1** [comment]
@burakksglu We are wondering if this logic, i.e. only allowing users with valid Grafana roles, could be enforced in the OAuth provider (Keycloak in your case)?

**Comment by burakksglu** [comment]
@aknuds1 as far as I know that's what we are doing actually, but this PR also needed for this logic to work. 

For example we have 2 grafana clients on keycloak, each with different users and instance. Let's say I have a role on the grafana A and I tried to login to grafana B. Without this PR I can login with "Auto Assigned Role Viewer" which is not wanted in our use case.

If I don't deny role and don't give any roles from Keycloak I still can login. To my knowledge OAuth does role mapping but it only maps and it doesn't have any role deny. 

Therefore this should be handled on the application side.

**Comment by aknuds1** [comment]
@burakksglu So basically you're sure that Keycloak can't deny certain users logging into a Grafana instance? As far as I can tell that is the crux of the matter; if Keycloak can deny the undesired users, the exclusion doesn't have to happen in Grafana.

**Comment by burakksglu** [comment]
> @burakksglu So basically you're sure that Keycloak can't deny certain users logging into a Grafana instance? As far as I can tell that is the crux of the matter; if Keycloak can deny the undesired users, the exclusion doesn't have to happen in Grafana.

No I don't think so. There are 2 other applications we test and they all do this in the application side.

**Comment by aknuds1** [comment]
[This SO question](https://stackoverflow.com/questions/54305880/how-can-i-restrict-client-access-to-only-one-group-of-users-in-keycloak/54374808#54374808) corresponds to your case right?

**Comment by burakksglu** [comment]
> [This SO question](https://stackoverflow.com/questions/54305880/how-can-i-restrict-client-access-to-only-one-group-of-users-in-keycloak/54374808#54374808) corresponds to your case right?

I'm going to check this out let you know. Thank you for forwarding this to me. I'll try to do analysis of it.

**Comment by burakksglu** [comment]
Hey @aknuds1,

We've tried and tested the given suggestions on the link you've provided. Unfortunately it doesn't work as many others stated in that topic. 

There are resource based policies that you can create on keycloak but in order to make it work both Grafana and Keycloak should have the support for it. For example Keycloak needs a resource server from Grafana, some configuration and js scripts.

After all it still might not work on the different configuration. This PR provides an application based role policy which most of the applications implemented in some way.

**Comment by levkar** [comment]
Currently, anyone on the realm becomes a valid user and there is no way to stop him from logging to grafana. Is there a simple way to deny login if user doesn't have a valid role?

**Comment by chelnak** [comment]
Yeah I can second this behaviour. It’s an issue when you share your IDP
with a wider org.

**Comment by aknuds1** [comment]
@marefr @bergquist @papagian - what are your thoughts? I don't have the most experience with OAuth.

> We've tried and tested the given suggestions on the link you've provided. Unfortunately it doesn't work as many others stated in that topic.
> 
> There are resource based policies that you can create on keycloak but in order to make it work both Grafana and Keycloak should have the support for it. For example Keycloak needs a resource server from Grafana, some configuration and js scripts.
> 
> After all it still might not work on the different configuration. This PR provides an application based role policy which most of the applications implemented in some way.

**Comment by bergquist** [comment]
I think this feature request is reasonable and the current implementation is fairly isolated. The OAuth code in Grafana is what it is....

But I'm not sure we only want this for generic OAuth. Why not for all OAuth providers?
The current implementation is fairly simple but since I can see this feature grow I would like to see some tests for `IsRoleAllowed`.

**Comment by aknuds1** [comment]
Thanks for giving your 2 cents @bergquist!

**Comment by levkar** [comment]
@bergquist 
it would require some effort to implement and test this in every single oauth provider supported. Some providers do not even have role capability. This PR prepares a base for this feature and it can be easily implemented in other providers as needed.

**Comment by bergquist** [comment]
Related https://github.com/grafana/grafana/issues/22741

**Comment by levkar** [comment]
we need your review & approval for this PR. thanks in advance!
@torkelo @daniellee @papagian @marefr @bergquist

**Comment by twendt** [comment]
Wouldn't the IsRoleAllowed function have the same implementation in all providers that support roles? It only checks if the role is one of the valid grafana roles.

What about having the function in social.go and instead have an IsRoleRequired function in all providers that returns the value of the roleRequired field.

During the creation of the SocialMap in social.go the value of roleRequired could then be hard coded to false for all providers that do not support roles.

I would also suggest to use the constants from the models package models.ROLE_ADMIN, models.ROLE_EDITOR and models.ROLE_VIEWER.

**Comment by aknuds1** [comment]
In order to fix the build-pipeline status issue, please merge with latest Grafana master.

**Comment by burakksglu** [comment]
@twendt : Thank you for the suggestions, I'll work on that package models and try to have the function on the social go. But just to be sure since I'm not expert on Go language, you are saying that we can have the "IsRoleAllowed" function on the social.go and override the return field based on the role availability in the oauth.go packages. I am going to work on that today.

@aknuds1 Ok, i did pulled from origin of master and tried to merge but it says I am up to date with master branch. So I am not sure about the pipeline issue. I also did git fetch.

**Comment by twendt** [comment]
@bergquist @marefr Could you please also comment on my suggestions?

**Comment by aknuds1** [comment]
@burakksglu According to GitHub, your branch is 228 commits behind Grafana master. Please try again merging Grafana master into your branch and push the changes.

**Comment by burakksglu** [comment]
So I've merged this branch with the master and worked on the suggestions you've made. Now the control function checks based on the models.ROLE_... constants. I have made the local test with keycloak, it's working as expected.

About the other suggestion, "hard coding the IsRoleAllowed in the social.go" I checked but couldn't figure it out. The way operation of this ' IsRoleAllowed ' function was similar to ' IsEmailAllowed '. We think that they work and should work in a similar way. If you look both of them they are actually constructed in similar way. 

Therefore I didn't make any changes on that side.

**Comment by aknuds1** [comment]
You don't need to make any more changes for now, we still need to make a decision internally with regards to including this functionality.

**Comment by burakksglu** [comment]
> You don't need to make any more changes for now, we still need to make a decision internally with regards to including this functionality.

OK, thank you for your effort and time. I am looking forward to see this functionality included in Grafana. Stay safe and healty in the mean time.

**Comment by levkar** [comment]
> You don't need to make any more changes for now, we still need to make a decision internally with regards to including this functionality.

thanks @aknuds1

**Comment by mr-yaky** [comment]
We have the same case with Keycloak and can't restring users from login if they not contain role for Grafana on Keycloak side. So, this feature will be also very useful for us.

**Comment by aknuds1** [comment]
Thank you for making the PR and the effort you've put into it, we really appreciate it. Unfortunately we've come to the conclusion that we will have to close it for now.                                                                                                                                                                                                                                                                                                                                                                 

We're not sure that adding this feature to only generic OAuth is a good thing to do at the moment. We would rather track and gather requirements for restricting users to login if certain criterias aren't met, with all different auth types in Grafana in mind.                                                                                          
                                                                                                                                                                              
Additionally, we think that this type of functionality crosses into enterprise edition territory, so we would like to determine for example if this functionality is actually useful outside of large companies.                                                                                                                                            
                                                                                                                                                                              
We'd like to suggest that you open a feature request for the desired functionality, linking to this PR, and ask for the community to vote on it, explaining their use case and adding their specific requirements. Just because we're closing it doesn't mean we cant reopen it later.

**Comment by mr-yaky** [comment]
>We would rather track and gather requirements for restricting users to login if certain criterias aren't met, with all different auth types in Grafana in mind.
@aknuds1 can we open an issue for this then please ? It supposed to be really useful feature!

**Comment by levkar** [comment]
Where do we open feature request? Issue?

**Comment by aknuds1** [comment]
@levkar @mr-yaky To make a feature request, simply create an issue in this repo, picking the "Enhancement request" template.

**Comment by burakksglu** [comment]
@mr-yaky @levkar @chelnak I've created the issue for Feature Request fyi: grafana/grafana/issues/23218

**Comment by levkar** [comment]
@chelnak @twendt @mr-yaky @aknuds1 please vote / thumbs up the request.

https://github.com/grafana/grafana/issues/23218

**Pull request closed without merging**
